haunt-roll-fail/arcs/game.scala

```diff
@@ -1,2728 +1,2744 @@
+                f.pay(x)
+
+                f.add(r)
+
+                u --> u.faction.reserve
+
+                f.log("released", u, "and took", r, x)
+
+                PressgangMainAction(f, AlreadyPaid, then)
+
+            case ExecuteMainAction(f, x, then) =>
+                implicit val convert = (u : Figure, selected : Boolean) => {
+                    val prefix = selected.not.??(u.faction.?./(_.short.toLowerCase + "-").|(""))
+                    val suffix = selected.??("-empty")
+
+                    Image(prefix + "agent" + suffix, styles.qship)
+                }
+
+                XXSelectObjectsAction(f, f.captives)
+                    .withGroup("Execute captives", x)
+                    .withThen(l => ExecuteAction(f, l, x, then))(l => ("Execute", l))
+                    .withExtras(CancelAction)
+
+            case ExecuteAction(f, l, x, then) =>
+                f.pay(x)
+
+                f.captives --> l --> f.trophies
+
+                f.log("executed", l, x)
+
+                then
+
+            case AbductMainAction(f, l, x, then) =>
+                Ask(f).group("Abduct".hl, "Agents", x)
+                    .each(market)(c => AbductAction(f, c, x, then).as(c).!(l.has(c).not))
+                    .cancel
+
+            case AbductAction(f, c, x, then) =>
+                f.pay(x)
+
+                val l = Influence(c).%(_.faction != f)
+
+                l --> f.captives
+
+                f.log("abducted", l, x)
+
+                then
+
+            case FarseersMainAction(f, e, then) =>
+                YYSelectObjectsAction(f, e.hand)
+                    .withGroup(Farseers, "look at", e, "cards")
+                    .withThen(d => FarseersAction(f, e, d, then))(d => ("Take", d, "from", e))("Take a card from", e)
+                    .withExtras(then.as("Skip"))
+
+            case FarseersAction(f, e, d, then) =>
+                e.hand --> d --> f.hand
+
+                f.log("took a card from", e, "with", Farseers)
+
+                YYSelectObjectsAction(f, f.hand)
+                    .withGroup(Farseers, "give back", e, "a card")
+                    .withThen(d => FarseersBackAction(f, e, d, then))(d => ("Give", d, "to", e))("Give a card back to", e)
+                    .withExtras(NoHand)
+
+            case FarseersBackAction(f, e, d, then) =>
+                f.hand --> d --> e.hand
+
+                f.log("gave back a card to", e)
+
+                then
+
+            case FarseersRedrawMainAction(f, then) =>
+                XXSelectObjectsAction(f, f.hand)
+                    .withGroup(Farseers, "discard cards to redraw")
+                    .withThen(l => FarseersRedrawAction(f, l, then))(l => ("Discard to draw", (l.num + 1).cards))
+                    .withExtras(NoHand, FarseersRedrawAction(f, $, then).as("Discard", Farseers, "to draw", 1.cards), CancelAction)
+
+            case FarseersRedrawAction(f, l, then) =>
+                val d = deck.num
```

Comment:
I think RAW cards are discarded by Farseers before they try to draw them, so the check to add open discard and shuffle should happen after.

Actually, this means that the only way Farseers can cause deck shuffle is if the deck is completely empty.
This might be impossible in a base 4p game. (Farseers -> Call to Action -> Guild Struggle -> Farseers -> Call to Action during the same round will make it empty, but Farseers will remain in discard.)
In a 3p game it is just highly improbable.

If no shuffle happens, Farseers will force redraw of the discarded cards if the deck doesn't have enough, in the order they were discarded. I'd say this means Farseers must be limited in UI to the number of cards in the deck, or if the deck is suddenly empty, to the number of cards in the open discard.

Reply:
`Call to Action` can't be reshuffled into the Court deck. Still, `Farseers` -> `Guild Struggle` -> `Call to Action` -> `Farseers` would exhaust a 3p deck on opening hand.

Reply:
Oh... `Guild Struggle` needs fixing then, it shuffles all cards into the deck right now, not just guild cards.